// @jefiozie/ngx-aws-deploy/src/lib/deploy/config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getCfDistributionId = exports.getSubFolder = exports.getRegion = exports.getBucket = exports.getSecretAccessKey = exports.getAccessKeyId = void 0;
const getAccessKeyId = () => {
    return process.env.NG_DEPLOY_AWS_ACCESS_KEY_ID || process.env.AWS_ACCESS_KEY_ID;
};
exports.getAccessKeyId = getAccessKeyId;
const getSecretAccessKey = () => {
    return process.env.NG_DEPLOY_AWS_SECRET_ACCESS_KEY || process.env.AWS_SECRET_ACCESS_KEY;
};
exports.getSecretAccessKey = getSecretAccessKey;
const getBucket = (builderConfig) => {
    return process.env.NG_DEPLOY_AWS_BUCKET || builderConfig.bucket;
};
exports.getBucket = getBucket;
const getRegion = (builderConfig) => {
    return process.env.NG_DEPLOY_AWS_REGION || builderConfig.region || process.env.AWS_DEFAULT_REGION;
};
exports.getRegion = getRegion;
const getSubFolder = (builderConfig) => {
    return process.env.NG_DEPLOY_AWS_SUB_FOLDER || builderConfig.subFolder;
};
exports.getSubFolder = getSubFolder;
const getCfDistributionId = (builderConfig) => {
    return process.env.NG_DEPLOY_AWS_CF_DISTRIBUTION_ID || builderConfig.cfDistributionId;
};
exports.getCfDistributionId = getCfDistributionId;
